### DISCUSSION TITLE
macOS: Mark window as dirty when it will require a prompt on close

### DISCUSSION THREAD

**Discussion by lilyball** [ghostty-org/ghostty#2678] [Feature Requests, Ideas] [answered]
macOS windows can be marked as "dirty", or edited. This is typically done in a document-based app to indicate that you've made changes to the document and that closing the window will prompt you to save your changes. This is displayed as a dot in the titlebar close button:

<img width="75" alt="screenshot" src="https://github.com/user-attachments/assets/f3b3f279-413a-4cd3-9186-22d3bc3af5fc">

It also adds an indicator to the corresponding menuitem in the Window menu.

Terminal.app uses this to indicate that the window/tab will prompt if you try and close it, and that's very useful. Besides just giving me a cue before I close it, it also means I can glance at the Window menu to see if any terminal windows are doing something (I do this before e.g. telling my computer to reboot, so I can deal with it before it interrupts the shutdown). iTerm also appears to be doing this too (I haven't checked to make sure it corresponds with the prompt-on-close condition but I'd be surprised if it doesn't).

I think it would be really useful if Ghostty picked up this behavior too. I don't know how to do this in SwiftUI but I'm pretty sure this is the `NSWindow.isDocumentEdited` property in AppKit.

**Comment by mitchellh** [comment]
I think this is generally useful as well.

> I don't know how to do this in SwiftUI but I'm pretty sure this is the NSWindow.isDocumentEdited property in AppKit.

Note that Ghostty _is_ an AppKit app, so we can use the AppKit API if we go forward with this. We embed SwiftUI views here and there, but it is not a SwiftUI App (the entry point is not SwiftUI).

**Reply by mitchellh** [reply, marked as answer]
#2723
